File: statsig-pyo3/py_src/statsig_python_core/statsig_types.py
import json
from typing import Any, Callable, Optional


def _log_error(tag: str, message: str):
    print(f"[Statsig::{tag}]: {message}")


class EvaluationDetails:
    reason: str
    lcut: Optional[int]
    received_at: Optional[int]
    version: Optional[int]

    def __init__(self, data: Optional[dict]):
        data = data or {}
        try:
            self.reason = data.get("reason") or ""
            self.lcut = data.get("lcut")
            self.received_at = data.get("received_at")
            self.version = data.get("version")
        except Exception as error:
            _log_error("EvaluationDetails", f"Failed to parse. {error}")

    def to_dict(self) -> dict:
        return {
            "reason": self.reason,
            "lcut": self.lcut,
            "received_at": self.received_at,
            "version": self.version,
        }


class BaseEvaluation:
    name: str
    rule_id: str
    id_type: str
    details: EvaluationDetails

    def __init__(self, name: str, data: dict):
        self.name = name
        self.rule_id = data.get("ruleID") or ""
        self.id_type = data.get("idType") or ""
        self.details = EvaluationDetails(data.get("details"))

    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "rule_id": self.rule_id,
            "id_type": self.id_type,
            "details": self.details.to_dict(),
        }

    def get_evaluation_details(self) -> EvaluationDetails:
        return self.details

    def get_name(self) -> str:
        return self.name

    def get_rule_id(self) -> str:
        return self.rule_id

    def get_id_type(self) -> str:
        return self.id_type

    def _get_typed(
        self,
        tag: str,
        value: dict,
        key: str,
        fallback: Any,
        exposure_func: Optional[Callable] = None,
    ) -> Any:
        res = value.get(key, None)
        if fallback is not None and not isinstance(fallback, type(res)):
            _log_error(
                tag,
                f"Type mismatch - '{self.name}.{key}'. Expected {type(fallback)}, got {type(res)}",
            )
            return fallback

        if res is not None:
            if exposure_func is not None:
                exposure_func(key)
            return res

        return fallback


class FeatureGate(BaseEvaluation):
    value: bool

    def __init__(self, name: str, raw: str):
        try:
            data = json.loads(raw) or {}
            super().__init__(name, data)
            self.value = data.get("value") or False
        except Exception as error:
            super().__init__(name, {})
            self.value = False
            _log_error("FeatureGate", f"Failed to parse. {error}")

    def to_dict(self) -> dict:
        base_dict = super().to_dict()
        base_dict["value"] = self.value
        return base_dict


class BaseConfigEvaluation(BaseEvaluation):
    value: dict
    __tag: str

    def __init__(self, name: str, data: dict, tag: str):
        self.__tag = tag
        super().__init__(name, data)
        self.value = data.get("value") or {}

    def get_value(self) -> dict:
        return self.value

    def get(self, param_name: str, fallback: Any = None) -> Any:
        return self._get_typed(self.__tag, self.value, param_name, fallback)

    def get_typed(self, param_name: str, fallback: Any = None) -> Any:
        return self._get_typed(self.__tag, self.value, param_name, fallback)

    def get_string(self, param_name: str, fallback: str) -> str:
        return self._get_typed(self.__tag, self.value, param_name, fallback)

    def get_integer(self, param_name: str, fallback: int) -> int:
        return self._get_typed(self.__tag, self.value, param_name, fallback)

    def get_float(self, param_name: str, fallback: float) -> float:
        actual_fallback = float(fallback) if isinstance(fallback, int) else fallback
        return self._get_typed(self.__tag, self.value, param_name, actual_fallback)

    def get_bool(self, param_name: str, fallback: bool) -> bool:
        return self._get_typed(self.__tag, self.value, param_name, fallback)

    def get_array(self, param_name: str, fallback: list) -> list:
        return self._get_typed(self.__tag, self.value, param_name, fallback)

    def get_object(self, param_name: str, fallback: dict) -> dict:
        return self._get_typed(self.__tag, self.value, param_name, fallback)

    def to_dict(self) -> dict:
        base_dict = super().to_dict()
        base_dict["value"] = self.value
        return base_dict


class DynamicConfig(BaseConfigEvaluation):
    def __init__(self, name: str, raw: str):
        try:
            data = json.loads(raw) or {}
            super().__init__(name, data, "DynamicConfig")
        except Exception as error:
            super().__init__(name, {}, "DynamicConfig")
            _log_error("DynamicConfig", f"Failed to parse. {error}")


class Experiment(BaseConfigEvaluation):
    group_name: Optional[str]

    def __init__(self, name: str, raw: str):
        try:
            data = json.loads(raw) or {}
            super().__init__(name, data, "Experiment")
            self.group_name = data.get("groupName")
        except Exception as error:
            super().__init__(name, {}, "Experiment")
            self.group_name = None
            _log_error("Experiment", f"Failed to parse. {error}")

    def to_dict(self) -> dict:
        base_dict = super().to_dict()
        base_dict["group_name"] = self.group_name
        return base_dict


class Layer(BaseEvaluation):
    group_name: Optional[str]
    allocated_experiment_name: Optional[str]
    __value: dict
    __exposure_func: Optional[Callable]

    def __init__(self, exposure_func: Callable, name: str, raw: str):
        try:
            data = json.loads(raw) or {}
            super().__init__(name, data)
            self.group_name = data.get("groupName")
            self.allocated_experiment_name = data.get("allocatedExperimentName")
            self.__value = data.get("value") or {}
            self.__exposure_func = exposure_func
        except Exception as error:
            super().__init__(name, {})
            self.group_name = None
            self.allocated_experiment_name = None
            self.__value = {}
            self.__exposure_func = None
            _log_error("Layer", f"Failed to parse. {error}")

    def get_value(self) -> dict:
        return self.__value

    def get(self, param_name: str, fallback: Any = None) -> Any:
        return self._get_typed(
            "Layer", self.__value, param_name, fallback, self._log_layer_param_exposure
        )

    def get_typed(self, param_name: str, fallback: Any = None) -> Any:
        return self._get_typed(
            "Layer", self.__value, param_name, fallback, self._log_layer_param_exposure
        )

    def get_string(self, param_name: str, fallback: str) -> str:
        return self._get_typed(
            "Layer", self.__value, param_name, fallback, self._log_layer_param_exposure
        )

    def get_integer(self, param_name: str, fallback: int) -> int:
        return self._get_typed(
            "Layer", self.__value, param_name, fallback, self._log_layer_param_exposure
        )

    def get_float(self, param_name: str, fallback: float) -> float:
        actual_fallback = float(fallback) if isinstance(fallback, int) else fallback
        return self._get_typed(
            "Layer",
            self.__value,
            param_name,
            actual_fallback,
            self._log_layer_param_exposure,
        )

    def get_bool(self, param_name: str, fallback: bool) -> bool:
        return self._get_typed(
            "Layer", self.__value, param_name, fallback, self._log_layer_param_exposure
        )

    def get_array(self, param_name: str, fallback: list) -> list:
        return self._get_typed(
            "Layer", self.__value, param_name, fallback, self._log_layer_param_exposure
        )

    def get_object(self, param_name: str, fallback: dict) -> dict:
        return self._get_typed(
            "Layer", self.__value, param_name, fallback, self._log_layer_param_exposure
        )

    def to_dict(self) -> dict:
        base_dict = super().to_dict()
        base_dict["group_name"] = self.group_name
        base_dict["allocated_experiment_name"] = self.allocated_experiment_name
        base_dict["__value"] = self.__value
        return base_dict

    def _log_layer_param_exposure(self, param_name: str):
        if self.__exposure_func is None:
            return

        self.__exposure_func(param_name)

File: statsig-node/src/lib/statsig_types.ts
import { EvaluationDetails, SecondaryExposure } from './statsig-generated';

export type TypedGet = <T = unknown>(
  key: string,
  defaultValue: T,
  typeGuard?: ((value: unknown) => value is T | null) | null,
) => T;

export type UnknownGet = (
  key: string,
  fallback?: boolean | number | string | object | Array<any> | null,
) => unknown | null;

function _getTypeOf(x: unknown): string {
  return Array.isArray(x) ? 'array' : x === null ? 'null' : typeof x;
}

class BaseEvaluation {
  readonly name: string;
  readonly ruleID: string;
  readonly idType: string;
  readonly details: EvaluationDetails;
  readonly secondaryExposures: SecondaryExposure[] = [];

  constructor(name: string, data: Record<string, unknown>) {
    this.name = name;
    this.ruleID = (data.ruleID as string) ?? '';
    this.idType = (data.idType as string) ?? '';
    this.details = _extractEvaluationDetails(
      data.details as Record<string, unknown>,
    );
    this.secondaryExposures =
      (data.secondaryExposures as SecondaryExposure[]) ?? [];
  }

  getEvaluationDetails(): EvaluationDetails {
    return this.details;
  }

  getRuleId(): string {
    return this.ruleID;
  }

  getIdType(): string {
    return this.idType;
  }

  getSecondaryExposures(): SecondaryExposure[] {
    return this.secondaryExposures;
  }
}

export class FeatureGate extends BaseEvaluation {
  readonly value: boolean;

  constructor(name: string, raw: string | null) {
    const data = _parseRawEvaluation(raw);

    super(name, data);

    this.value = data.value === true;
  }
}

export class DynamicConfig extends BaseEvaluation {
  readonly value: Record<string, unknown> = {};

  readonly get: TypedGet;
  readonly getValue: UnknownGet;

  constructor(name: string, raw: string) {
    const data = _parseRawEvaluation(raw);

    super(name, data);

    this.value = (data.value as Record<string, unknown>) ?? {};

    this.get = _makeTypedGet(name, this.value);
    this.getValue = _makeUnknownGet(this.value);
  }
}

export class Experiment extends BaseEvaluation {
  readonly groupName: string | null = null;
  readonly value: Record<string, unknown> = {};

  readonly get: TypedGet;
  readonly getValue: UnknownGet;

  constructor(name: string, raw: string) {
    const data = _parseRawEvaluation(raw);

    super(name, data);

    this.groupName = (data.groupName as string) ?? null;
    this.value = (data.value as Record<string, unknown>) ?? {};

    this.get = _makeTypedGet(name, this.value);
    this.getValue = _makeUnknownGet(this.value);
  }

  getGroupName(): string | null {
    return this.groupName;
  }
}

export class Layer extends BaseEvaluation {
  readonly groupName: string | null = null;
  readonly allocatedExperimentName: string | null = null;
  readonly __value: Record<string, unknown> = {};

  readonly get: TypedGet;
  readonly getValue: UnknownGet;

  constructor(exposeFn: (param: string) => void, name: string, raw: string) {
    const data = _parseRawEvaluation(raw);

    super(name, data);

    this.__value = (data.value as Record<string, unknown>) ?? {};
    this.groupName = (data.groupName as string) ?? null;
    this.allocatedExperimentName =
      (data.allocatedExperimentName as string) ?? null;

    this.get = _makeTypedGet(name, this.__value, exposeFn);
    this.getValue = _makeUnknownGet(this.__value, exposeFn);
  }

  getGroupName(): string | null {
    return this.groupName;
  }

  getAllocatedExperimentName(): string | null {
    return this.allocatedExperimentName;
  }
}

function _parseRawEvaluation(raw: string | null): Record<string, unknown> {
  try {
    return JSON.parse(raw ?? '{}') as Record<string, unknown>;
  } catch (error) {
    console.error(`[Statsig] Error parsing BaseEvaluation: ${error}`);
    return {};
  }
}

function _makeTypedGet(
  name: string,
  value: Record<string, unknown>,
  exposeFunc?: (param: string) => void,
): TypedGet {
  return <T = unknown>(
    key: string,
    defaultValue: T,
    typeGuard: ((value: unknown) => value is T | null) | null = null,
  ): T => {
    // @ts-ignore - intentionally matches legacy behavior exactly
    defaultValue = defaultValue ?? null;

    // Equivalent to legacy `this.getValue(key, defaultValue)`
    const val = (value[key] ?? defaultValue) as unknown;

    if (val == null) {
      return defaultValue;
    }

    const expectedType = _getTypeOf(defaultValue);
    const actualType = _getTypeOf(val);

    if (typeGuard != null) {
      if (typeGuard(val)) {
        exposeFunc?.(key);
        return val as T;
      }

      console.warn(
        `[Statsig] Parameter type mismatch. '${name}.${key}' failed typeGuard. Expected '${expectedType}', got '${actualType}'`,
      );
      return defaultValue;
    }

    if (defaultValue == null || expectedType === actualType) {
      exposeFunc?.(key);
      return val as T;
    }

    console.warn(
      `[Statsig] Parameter type mismatch. '${name}.${key}' was found to be type '${actualType}' but fallback/return type is '${expectedType}'`,
    );
    return defaultValue;
  };
}

function _makeUnknownGet(
  value: Record<string, unknown>,
  exposeFunc?: (param: string) => void,
): UnknownGet {
  return (param: string, fallback?: unknown) => {
    if (fallback === undefined) {
      fallback = null;
    }

    if (param == null) {
      return fallback;
    }

    if (value[param] != null) {
      exposeFunc?.(param);
    }

    return value[param] ?? fallback;
  };
}

function _extractEvaluationDetails(
  data: Record<string, unknown> | null,
): EvaluationDetails {
  if (data == null) {
    return {
      reason: '',
      lcut: 0,
      receivedAt: 0,
      version: 0,
    };
  }
  return {
    reason: data.reason as string,
    lcut: data.lcut as number,
    receivedAt: data.received_at as number,
    version: data.version as number,
  };
}
